tests/ssl/mutual_authentication_test.py

```diff
@@ -0,0 +1,137 @@
+        cluster = self.create_cluster(self.rc, self.configure_cluster(True))
+        member = cluster.start_member()
+
+        with self.assertRaises(HazelcastError):
+            client = HazelcastClient(get_ssl_config(True,
+                                                    get_abs_path(self.current_directory, "server1-cert.pem"),
+                                                    get_abs_path(self.current_directory, "client2-cert.pem"),
+                                                    get_abs_path(self.current_directory, "client2-key.pem"),
+                                                    protocol=PROTOCOL.TLSv1))
+
+    def test_required_client_and_server_not_authenticated(self):
+        cluster = self.create_cluster(self.rc, self.configure_cluster(True))
+        member = cluster.start_member()
+
+        with self.assertRaises(HazelcastError):
+            client = HazelcastClient(get_ssl_config(True,
+                                                    get_abs_path(self.current_directory, "server2-cert.pem"),
+                                                    get_abs_path(self.current_directory, "client2-cert.pem"),
+                                                    get_abs_path(self.current_directory, "client2-key.pem"),
+                                                    protocol=PROTOCOL.TLSv1))
+
+    def test_optional_client_and_server_authenticated(self):
+        cluster = self.create_cluster(self.rc, self.configure_cluster(False))
+        member = cluster.start_member()
+        client = HazelcastClient(get_ssl_config(True,
+                                                get_abs_path(self.current_directory, "server1-cert.pem"),
+                                                get_abs_path(self.current_directory, "client1-cert.pem"),
+                                                get_abs_path(self.current_directory, "client1-key.pem"),
+                                                protocol=PROTOCOL.TLSv1))
+        self.assertTrue(client.lifecycle.is_live)
+        client.shutdown()
+
+    def test_optional_server_not_authenticated(self):
+        cluster = self.create_cluster(self.rc, self.configure_cluster(False))
+        member = cluster.start_member()
+
+        with self.assertRaises(HazelcastError):
+            client = HazelcastClient(get_ssl_config(True,
+                                                    get_abs_path(self.current_directory, "server2-cert.pem"),
+                                                    get_abs_path(self.current_directory, "client1-cert.pem"),
+                                                    get_abs_path(self.current_directory, "client1-key.pem"),
+                                                    protocol=PROTOCOL.TLSv1))
+
+    def test_optional_client_not_authenticated(self):
+        cluster = self.create_cluster(self.rc, self.configure_cluster(False))
+        member = cluster.start_member()
+
+        with self.assertRaises(HazelcastError):
```

Comment:
If mutual authentication is optional why shouldnt this one connect?

Reply:
mutual authentication is optional but we still set a client certificate. In that case, server will try to authenticate client with this invalid certificates and it will fail